src/sys/stat.rs

```diff
@@ -7,14 +6,357 @@ pub use libc::c_uint;
+    any(target_arch = "arm", target_arch = "armv7", target_arch = "x86")
+))]
+use std::convert::TryInto;
+use std::{mem, os::unix::io::RawFd};
+
+#[repr(transparent)]
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+/// File status, a wrapper type for `libc::stat`
+pub struct FileStat(libc::stat);
+
+impl FileStat {
+    /// Constructs a new `FileStat` from raw `libc::stat`
+    pub fn from_raw(raw_stat: libc::stat) -> Self {
+        Self(raw_stat)
+    }
+
+    /// Gets the raw `libc::stat` wrapped by self.
+    pub fn as_raw(&self) -> libc::stat {
+        self.0
+    }
+
+    #[cfg(all(
+        target_os = "android",
+        any(target_arch = "arm", target_arch = "armv7", target_arch = "x86")
+    ))]
+    /// Returns the ID of device containing this file
+    pub fn dev(&self) -> libc::c_ulonglong {
+        self.0.st_dev
+    }
+
+    #[cfg(all(
+        target_os = "linux",
+        any(target_arch = "mips", target_arch = "mipsel")
+    ))]
+    /// Returns the ID of device containing this file
+    pub fn dev(&self) -> libc::c_ulong {
+        self.0.st_dev
+    }
+
+    #[cfg(not(any(
+        all(
+            target_os = "android",
+            any(
+                target_arch = "arm",
+                target_arch = "armv7",
+                target_arch = "x86"
+            )
+        ),
+        all(
+            target_os = "linux",
+            any(target_arch = "mips", target_arch = "mipsel")
+        )
+    )))]
+    /// Returns the ID of device containing this file
+    pub fn dev(&self) -> dev_t {
+        self.0.st_dev
+    }
+
+    #[cfg(all(
+        target_os = "android",
+        any(target_arch = "arm", target_arch = "armv7", target_arch = "x86")
+    ))]
+    /// Returns the Inode number
+    pub fn ino(&self) -> libc::c_ulonglong {
+        self.0.st_ino
+    }
+
+    #[cfg(not(all(
+        target_os = "android",
+        any(target_arch = "arm", target_arch = "armv7", target_arch = "x86")
+    )))]
+    /// Returns the Inode number
+    pub fn ino(&self) -> ino_t {
+        self.0.st_ino
+    }
+
+    #[cfg(all(target_os = "android", target_arch = "x86_64"))]
+    /// Returns the number of hard links
+    pub fn nlink(&self) -> libc::c_ulong {
+        self.0.st_nlink
+    }
+
+    #[cfg(not(all(target_os = "android", target_arch = "x86_64")))]
+    /// Returns the number of hard links
+    pub fn nlink(&self) -> nlink_t {
+        self.0.st_nlink
+    }
+
+    #[cfg(all(
+        target_os = "android",
+        any(target_arch = "arm", target_arch = "armv7", target_arch = "x86")
+    ))]
+    /// Returns a number encoding `file type` and `mode`
+    ///
+    /// We have dedicated types for `file type`
+    /// ([SFlag](https://docs.rs/nix/latest/nix/sys/stat/struct.SFlag.html)) and
+    /// `mode` ([Mode](https://docs.rs/nix/latest/nix/sys/stat/struct.Mode.html)),
+    /// You should use them instead of using the raw numeric type.
+    pub fn mode(&self) -> mode_t {
+        self.0.st_mode.try_into().unwrap()
```

Comment:
On `Android: arm/armv7/i686`, the type of `st_mode` is `c_uin`
```rust
st_mode: c_uint
```
And `mode_t` is defined as `u16`
```rust
pub type mode_t = u16;
```
And the constants of `Mode` and `SFlag` are defined as `mode_t`
```rust
...
pub const S_IFIFO: mode_t = 4096;
pub const S_IRWXU: mode_t = 448;
...
```

So I make this function return `mode_t`.

Perhaps I should replace `.unwrap()` with `.expect("BUG: st_mode returns an invalid number")`

Reply:
Hello, I'm from https://github.com/andreacorbellini/rust-atomic-write-file/issues/9 and encountered the same problem. Is there any further information?

Reply:
replied there.